x/stake/keeper/validator.go

```diff
@@ -693,6 +693,37 @@ func (k Keeper) RemoveValidator(ctx sdk.Context, address sdk.ValAddress) {
 	tstore.Set(GetTendermintUpdatesTKey(address), bz)
 }
 
+// UpdateValidatorCommission attempts to update a validator's commission rate.
+// An error is returned if the new commission rate is invalid.
+func (k Keeper) UpdateValidatorCommission(ctx sdk.Context, validator types.Validator, newRate sdk.Dec) sdk.Error {
+	commission := validator.Commission
+	blockTime := ctx.BlockHeader().Time
+
+	switch {
+	case blockTime.Sub(validator.CommissionUpdateTime).Hours() < 24:
+		// new rate cannot be changed more than once within 24 hours
+		return types.ErrCommissionUpdateTime(k.Codespace())
+
+	case newRate.LT(sdk.ZeroDec()):
+		// new rate cannot be negative
+		return types.ErrCommissionNegative(k.Codespace())
+
+	case newRate.GT(commission.MaxRate):
+		// new rate cannot be greater than the max rate
+		return types.ErrCommissionGTMaxRate(k.Codespace())
+
+	case newRate.Sub(commission.Rate).Abs().GT(commission.MaxChangeRate):
+		// new rate % change cannot be greater than the max change rate
+		return types.ErrCommissionGTMaxChangeRate(k.Codespace())
+	}
```

Comment:
I really like the approach that was taken to separate out the commission logic to a new commission type, However the commission type has much dup logic to what is here in `UpdateValidatorCommission` - I know one the reasons this check can't simply call `commission.Validate()` is because of the time check on `validator.CommissionUpdateTime`

SO what I'd suggest is that `CommissionUpdateTime` be moved into the commission object (which I think is categorically more correct than adding it directly to the validator object). This way we can have ALL the error checking and update logic in the `commission.go` removing all the dup logic from `UpdateValidatorCommission` (which then should only be a few lines of code)

Reply:
On further review, I think I see what you were trying to do here, reuse the `Commission` object for Msg/Tx types - and for that input of commission you don't need the `CommissionUpdateTime` field. -> I still think it makes sense to add this field back to the commission object to consolidate the logic, this would mean now that we would need to change the `Msg/Tx` types to not using the commission object anymore but just using the 3 relevant fields from commission - I think that's totally fine though

Reply:
Indeed you followed my flow of reasoning correctly. I will amend to have `CommissionUpdateTime` be part of the commission object and update the messages/handlers accordingly.